source/core/slang-offset-container.h

```diff
@@ -0,0 +1,359 @@
+    }
+
+    // Now lets write to it
+    {
+        // We can have raw pointer (or reference) to a thing but we need to be *careful* if we allocate 
+        Thing* rawThing = base[thing];
+        // We are okay here, nothing between getting the raw pointer and the write allocated/newed anything on the OffsetContainer
+        rawThing->value = 20;
+
+        // Lets set up name 
+        Offset32Ptr<OffsetString> text = offsetContainer.newString("Hello World!");
+
+        // BAD! The rawThing point could now be invalid because the call to newString may have had to allocate more memory
+        rawThing->text = text;
+
+        // This is okay 
+        base[thing]->text = text;
+
+        // Or we can update rawThing such that is up to date
+        rawThing = base[thing];
+        // So now this is okay again
+        rawThing->text = text;
+    }
+}
+
+```
+*/
+
+enum
+{
+    kNull32Offset = 0,                                  
+    kStartOffset = uint32_t(sizeof(uint64_t)),          ///< The offset to the first contained thing
+};
+
+template <typename T>
+class Offset32Ref;
+
+/* A pointer to items held in OffsetContainer (or OffsetBase relative) that remains correct even if
+the memory inside OffsetContainer moves.
+*/
+template <typename T>
+class Offset32Ptr
+{
+public:
+    typedef Offset32Ptr ThisType;
+
+    const ThisType& operator=(const ThisType& rhs) { m_offset = rhs.m_offset; return *this; }
+    bool operator==(const ThisType& rhs) const { return m_offset == rhs.m_offset; }
+    bool operator!=(const ThisType& rhs) const { return m_offset != rhs.m_offset; }
+
+    bool operator<(const ThisType& rhs) const { return m_offset < rhs.m_offset; }
+    bool operator<=(const ThisType& rhs) const { return m_offset <= rhs.m_offset; }
+    bool operator>(const ThisType& rhs) const { return m_offset > rhs.m_offset; }
+    bool operator>=(const ThisType& rhs) const { return m_offset >= rhs.m_offset; }
+
+    operator bool() const { return m_offset != kNull32Offset; }
+
+    Offset32Ref<T> operator*();
+
+    ThisType& operator++() { m_offset += uint32_t(sizeof(T)); return *this; }
+    ThisType operator++(int) { const auto offset = m_offset; m_offset += uint32_t(sizeof(T)); return ThisType(offset); }
+
+    ThisType& operator--() { m_offset -= sizeof(T); return *this; }
+    ThisType operator--(int) { const auto offset = m_offset; m_offset -= uint32_t(sizeof(T)); return ThisType(offset); }
+
+    friend ThisType operator+(const ThisType& a, Index b) { return ThisType(a.m_offset + uint32_t(sizeof(T) * b)); }
+    friend ThisType operator+(Index a, const ThisType& b) { return ThisType(b.m_offset + uint32_t(sizeof(T) * a)); }
+
+    bool isNull() const { return m_offset == kNull32Offset; }
+
+    void setNull() { m_offset = kNull32Offset; }
+    Offset32Ptr():m_offset(kNull32Offset) {}
+    Offset32Ptr(const ThisType& rhs): m_offset(rhs.m_offset) {}
+    explicit Offset32Ptr(uint32_t offset): m_offset(offset) {}
+
+    uint32_t m_offset;
```

Comment:
Is it intentional that `m_offset` is `public`? I don't have a strong opinion, but I'm a bit surprised.

Reply:
Yes, arguably it shouldn't be.